### ISSUE TITLE
[Bug]: [benchmark] diskann builds indexes too slowly for more than 16h after inserting 100 million data

### ISSUE THREAD

**Issue by elstic** [milvus-io/milvus#25891]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version:2.2.0-20230721-b2a0aec9
- Deployment mode(standalone or cluster):cluster
- MQ type(rocksmq, pulsar or kafka):    pulsar
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

argo task : fouramf-fnbw2
case: test_concurrent_locust_100m_diskann_ddl_dql_filter_cluster

server:
```
fouramf-fnbw2-23-6813-etcd-0                                      1/1     Running            0              6m      10.104.21.91    4am-node24   <none>           <none>
fouramf-fnbw2-23-6813-etcd-1                                      1/1     Running            0              6m      10.104.23.116   4am-node27   <none>           <none>
fouramf-fnbw2-23-6813-etcd-2                                      1/1     Running            0              6m      10.104.24.215   4am-node29   <none>           <none>
fouramf-fnbw2-23-6813-milvus-datacoord-6c648c44b8-tqcpq           1/1     Running            0              6m      10.104.17.241   4am-node23   <none>           <none>
fouramf-fnbw2-23-6813-milvus-datanode-57466d7f97-4gnvf            1/1     Running            1 (119s ago)   6m      10.104.4.124    4am-node11   <none>           <none>
fouramf-fnbw2-23-6813-milvus-indexcoord-77f6bb4f84-mx5mm          1/1     Running            1 (2m ago)     6m      10.104.18.114   4am-node25   <none>           <none>
fouramf-fnbw2-23-6813-milvus-indexnode-67ccdf574f-6fksn           1/1     Running            0              6m      10.104.21.87    4am-node24   <none>           <none>
fouramf-fnbw2-23-6813-milvus-proxy-566895978-9znv4                1/1     Running            1 (119s ago)   6m      10.104.21.86    4am-node24   <none>           <none>
fouramf-fnbw2-23-6813-milvus-querycoord-5d69df9f5f-t6t2x          1/1     Running            1 (119s ago)   6m      10.104.4.116    4am-node11   <none>           <none>
fouramf-fnbw2-23-6813-milvus-querynode-67f9d6f699-gcgwc           1/1     Running            0              6m      10.104.17.242   4am-node23   <none>           <none>
fouramf-fnbw2-23-6813-milvus-rootcoord-7b49dd5b8f-6n5lh           1/1     Running            1 (2m ago)     6m      10.104.18.113   4am-node25   <none>           <none>
fouramf-fnbw2-23-6813-minio-0                                     1/1     Running            0              6m      10.104.21.93    4am-node24   <none>           <none>
fouramf-fnbw2-23-6813-minio-1                                     1/1     Running            0              6m      10.104.13.252   4am-node16   <none>           <none>
fouramf-fnbw2-23-6813-minio-2                                     1/1     Running            0              6m      10.104.9.115    4am-node14   <none>           <none>
fouramf-fnbw2-23-6813-minio-3                                     1/1     Running            0              5m59s   10.104.24.217   4am-node29   <none>           <none>
fouramf-fnbw2-23-6813-pulsar-bookie-0                             1/1     Running            0              6m      10.104.13.253   4am-node16   <none>           <none>
fouramf-fnbw2-23-6813-pulsar-bookie-1                             1/1     Running            0              6m      10.104.16.179   4am-node21   <none>           <none>
fouramf-fnbw2-23-6813-pulsar-bookie-2                             1/1     Running            0              5m59s   10.104.9.118    4am-node14   <none>           <none>
fouramf-fnbw2-23-6813-pulsar-bookie-init-dqhhm                    0/1     Completed          0              6m      10.104.17.240   4am-node23   <none>           <none>
fouramf-fnbw2-23-6813-pulsar-broker-0                             1/1     Running            0              6m      10.104.18.115   4am-node25   <none>           <none>
fouramf-fnbw2-23-6813-pulsar-proxy-0                              1/1     Running            0              6m      10.104.20.111   4am-node22   <none>           <none>
fouramf-fnbw2-23-6813-pulsar-pulsar-init-ps6ps                    0/1     Completed          0              6m      10.104.21.85    4am-node24   <none>           <none>
fouramf-fnbw2-23-6813-pulsar-recovery-0                           1/1     Running            0              6m      10.104.18.116   4am-node25   <none>           <none>
fouramf-fnbw2-23-6813-pulsar-zookeeper-0                          1/1     Running            0              6m      10.104.21.92    4am-node24   <none>           <none>
fouramf-fnbw2-23-6813-pulsar-zookeeper-1                          1/1     Running            0              5m19s   10.104.13.2     4am-node16   <none>           <none>
fouramf-fnbw2-23-6813-pulsar-zookeeper-2                          1/1     Running            0              4m43s   10.104.4.126    4am-node11   <none>           <none>
```

client log  （build index DISKANN: 59258.0836s）:
```
[2023-07-21 11:47:33,788 -  INFO - fouram]: [Time] Collection.insert run in 1.7031s (api_request.py:45)
[2023-07-21 11:47:33,791 -  INFO - fouram]: [Base] Number of vectors in the collection(fouram_WoV6ubON): 99900000 (base.py:483)
[2023-07-21 11:47:33,930 -  INFO - fouram]: [Base] Start inserting, ids: 99950000 - 99999999, data size: 100,000,000 (base.py:323)
[2023-07-21 11:47:35,823 -  INFO - fouram]: [Time] Collection.insert run in 1.8919s (api_request.py:45)
[2023-07-21 11:47:35,825 -  INFO - fouram]: [Base] Number of vectors in the collection(fouram_WoV6ubON): 99900000 (base.py:483)
[2023-07-21 11:47:35,881 -  INFO - fouram]: [Base] Total time of insert: 3395.2466s, average number of vector bars inserted per second: 29452.9417, average time to insert 50000 vectors per time: 1.6976s (base.py:394)
[2023-07-21 11:47:35,881 -  INFO - fouram]: [Base] Start flush collection fouram_WoV6ubON (base.py:292)
[2023-07-21 11:47:38,433 -  INFO - fouram]: [Base] Params of index: [{'float_vector': {'index_type': 'DISKANN', 'metric_type': 'L2', 'params': {}}}] (base.py:456)
[2023-07-21 11:47:38,433 -  INFO - fouram]: [Base] Start release collection fouram_WoV6ubON (base.py:303)
[2023-07-21 11:47:38,437 -  INFO - fouram]: [Base] Start build index of DISKANN for collection fouram_WoV6ubON, params:{'index_type': 'DISKANN', 'metric_type': 'L2', 'params': {}} (base.py:442)
[2023-07-22 04:15:16,522 -  INFO - fouram]: [Time] Index run in 59258.0836s (api_request.py:45)
[2023-07-22 04:15:16,523 -  INFO - fouram]: [CommonCases] RT of build index DISKANN: 59258.0836s (common_cases.py:96)
[2023-07-22 04:15:16,542 -  INFO - fouram]: [Base] Params of index: [{'float_vector': {'index_type': 'DISKANN', 'metric_type': 'L2', 'params': {}}}] (base.py:456)
[2023-07-22 04:15:16,542 -  INFO - fouram]: [CommonCases] Prepare index DISKANN done. (common_cases.py:99)
[2023-07-22 04:15:16,542 -  INFO - fouram]: [CommonCases] No scalars need to be indexed. (common_cases.py:107)
[2023-07-22 04:15:16,545 -  INFO - fouram]: [Base] Number of vectors in the collection(fouram_WoV6ubON): 100000000 (base.py:483)
[2023-07-22 04:15:16,545 -  INFO - fouram]: [Base] Start load collection fouram_WoV6ubON,replica_number:1,kwargs:{} (base.py:298)
[2023-07-22 04:52:08,089 -  INFO - fouram]: [Time] Collection.load run in 2211.543s (api_request.py:45)
```

### Expected Behavior

Insert 100 million data, diskann index build index time should be between 4 to 6 hours

For example, v2.2.10 build index takes time: 18407.8551s

### Steps To Reproduce

```markdown
1. create a collection or use an existing collection
        2. build index on vector column
        3. insert a certain number of vectors
        4. flush collection
        5. build index on vector column with the same parameters
        6. build index on on scalars column or not
        7. count the total number of rows
        8. load collection
        9. perform concurrent operations
        10. clean all collections or not
```

### Anything else?

Examples of what is still currently available: 
```
fouramf-mcjxs-39-4317-etcd-0                                      1/1     Running            0               16h     10.104.17.16    4am-node23   <none>           <none>
fouramf-mcjxs-39-4317-etcd-1                                      1/1     Running            0               16h     10.104.21.97    4am-node24   <none>           <none>
fouramf-mcjxs-39-4317-etcd-2                                      1/1     Running            0               16h     10.104.6.228    4am-node13   <none>           <none>
fouramf-mcjxs-39-4317-milvus-datacoord-857cbdd79-xhfks            1/1     Running            0               16h     10.104.21.84    4am-node24   <none>           <none>
fouramf-mcjxs-39-4317-milvus-datanode-5f7b45b7dd-cnlcn            1/1     Running            0               16h     10.104.5.173    4am-node12   <none>           <none>
fouramf-mcjxs-39-4317-milvus-indexcoord-6ff8c7769-swkf5           1/1     Running            0               16h     10.104.21.94    4am-node24   <none>           <none>
fouramf-mcjxs-39-4317-milvus-indexnode-7cf8669448-nt77f           1/1     Running            0               16h     10.104.16.191   4am-node21   <none>           <none>
fouramf-mcjxs-39-4317-milvus-proxy-6885d47777-grbfz               1/1     Running            0               16h     10.104.21.85    4am-node24   <none>           <none>
fouramf-mcjxs-39-4317-milvus-querycoord-7c95f66f8-74954           1/1     Running            0               16h     10.104.21.89    4am-node24   <none>           <none>
fouramf-mcjxs-39-4317-milvus-querynode-68bfb49bc4-bsrf7           1/1     Running            0               16h     10.104.18.175   4am-node25   <none>           <none>
fouramf-mcjxs-39-4317-milvus-rootcoord-7b67f9db8d-d44qg           1/1     Running            0               16h     10.104.21.88    4am-node24   <none>           <none>
fouramf-mcjxs-39-4317-minio-0                                     1/1     Running            0               16h     10.104.5.175    4am-node12   <none>           <none>
fouramf-mcjxs-39-4317-minio-1                                     1/1     Running            0               16h     10.104.17.15    4am-node23   <none>           <none>
fouramf-mcjxs-39-4317-minio-2                                     1/1     Running            0               16h     10.104.6.227    4am-node13   <none>           <none>
fouramf-mcjxs-39-4317-minio-3                                     1/1     Running            0               16h     10.104.4.10     4am-node11   <none>           <none>
fouramf-mcjxs-39-4317-pulsar-bookie-0                             1/1     Running            0               16h     10.104.12.144   4am-node17   <none>           <none>
fouramf-mcjxs-39-4317-pulsar-bookie-1                             1/1     Running            0               16h     10.104.9.20     4am-node14   <none>           <none>
fouramf-mcjxs-39-4317-pulsar-bookie-2                             1/1     Running            0               16h     10.104.17.21    4am-node23   <none>           <none>
fouramf-mcjxs-39-4317-pulsar-bookie-init-zkn79                    0/1     Completed          0               16h     10.104.16.190   4am-node21   <none>           <none>
fouramf-mcjxs-39-4317-pulsar-broker-0                             1/1     Running            0               16h     10.104.21.95    4am-node24   <none>           <none>
fouramf-mcjxs-39-4317-pulsar-proxy-0                              1/1     Running            0               16h     10.104.21.90    4am-node24   <none>           <none>
fouramf-mcjxs-39-4317-pulsar-pulsar-init-gpzsk                    0/1     Completed          0               16h     10.104.5.172    4am-node12   <none>           <none>
fouramf-mcjxs-39-4317-pulsar-recovery-0                           1/1     Running            0               16h     10.104.6.224    4am-node13   <none>           <none>
fouramf-mcjxs-39-4317-pulsar-zookeeper-0                          1/1     Running            0               16h     10.104.17.18    4am-node23   <none>           <none>
fouramf-mcjxs-39-4317-pulsar-zookeeper-1                          1/1     Running            0               16h     10.104.1.16     4am-node10   <none>           <none>
fouramf-mcjxs-39-4317-pulsar-zookeeper-2                          1/1     Running            0               16h     10.104.9.22     4am-node14   <none>           <none>
```

The build index is still pending after more than 12h： 
![image](https://github.com/milvus-io/milvus/assets/48523564/81d29c94-561e-4871-8170-a3a2521101a1)

**Comment by yanliang567** [comment]
discussed with @xiaocai2333 , we doubt that it was caused by
1.  all the parent-segments must be indexed before compaction 
2. and all the child-segments must be indexed before create_index completed.
We think we should improve this, to shorten the index time for users.

/assign @wayblink @xiaocai2333 
/unassign

**Comment by xiaofan-luan** [comment]
> discussed with @xiaocai2333 , we doubt that it was caused by
> 
> 1. all the parent-segments must be indexed before compaction
> 2. and all the child-segments must be indexed before create_index completed.
>    We think we should improve this, to shorten the index time for users.
> 
> /assign @wayblink @xiaocai2333 /unassign

think of optimize compaction rather than change the rule would be great

**Comment by xiaocai2333** [comment]
> 

Yes, we need better strategies to prevent invalid compactions.
I thought of a strategy:
- if the size of the target segment in the plan is less than 0.9(or another suitable ratio) of `segmentMaxSize`  and the number of segments is less than 4, do not execute the compaction.

**Comment by wayblink** [comment]
Thinking about middle size segments, I think only 2 size level are needed in common write path: maxSize*sealPortion,maxSize.

Middle size segments can only be created when delete, stale or frequently manual flush(too many small segments) happens.

And I think we can do some protection on manual flush. Don't actually execute flush if user call flush too frequently.

**Comment by XuanYang-cn** [comment]
Yes, way too slow. It takes 5h to insert and index 10M 768 Data, Now it takes 7hrs **JUST** index but still have 51(or even more) index tasks unissued.
![image](https://github.com/milvus-io/milvus/assets/51370125/2397a4a0-924a-4277-94ac-200464fba3bd)
![image](https://github.com/milvus-io/milvus/assets/51370125/f241b305-a322-4836-a339-bfba36d0b13c)

**Comment by XuanYang-cn** [comment]
![image](https://github.com/milvus-io/milvus/assets/51370125/3a517ed4-9625-43c4-91e1-cf577d90a15f)

**Comment by elstic** [comment]
This issue still exists, verify image: 2.2.0-20230807-ef31fe23

build index exceeded 18h and still not completed

server:
```
fouramf-bzvqc-73-2189-etcd-0                                      1/1     Running     0               19h     10.104.1.58     4am-node10   <none>           <none>
fouramf-bzvqc-73-2189-etcd-1                                      1/1     Running     0               19h     10.104.17.17    4am-node23   <none>           <none>
fouramf-bzvqc-73-2189-etcd-2                                      1/1     Running     0               19h     10.104.23.165   4am-node27   <none>           <none>
fouramf-bzvqc-73-2189-milvus-datacoord-7c44b68c9c-wqdj2           1/1     Running     0               19h     10.104.24.66    4am-node29   <none>           <none>
fouramf-bzvqc-73-2189-milvus-datanode-77b6dd8cd-s2zls             1/1     Running     1 (19h ago)     19h     10.104.24.70    4am-node29   <none>           <none>
fouramf-bzvqc-73-2189-milvus-indexcoord-5dcb54fc55-rm697          1/1     Running     1 (19h ago)     19h     10.104.6.61     4am-node13   <none>           <none>
fouramf-bzvqc-73-2189-milvus-indexnode-5685f69f86-fxgpc           1/1     Running     0               19h     10.104.13.56    4am-node16   <none>           <none>
fouramf-bzvqc-73-2189-milvus-proxy-594bff6567-skcss               1/1     Running     1 (19h ago)     19h     10.104.24.69    4am-node29   <none>           <none>
fouramf-bzvqc-73-2189-milvus-querycoord-58d4f47cbd-vjpd7          1/1     Running     1 (19h ago)     19h     10.104.24.67    4am-node29   <none>           <none>
fouramf-bzvqc-73-2189-milvus-querynode-697684c67b-6htwl           1/1     Running     0               19h     10.104.20.227   4am-node22   <none>           <none>
fouramf-bzvqc-73-2189-milvus-rootcoord-7547564c94-xlg2b           1/1     Running     1 (19h ago)     19h     10.104.15.81    4am-node20   <none>           <none>
fouramf-bzvqc-73-2189-minio-0                                     1/1     Running     0               19h     10.104.1.53     4am-node10   <none>           <none>
fouramf-bzvqc-73-2189-minio-1                                     1/1     Running     0               19h     10.104.14.135   4am-node18   <none>           <none>
fouramf-bzvqc-73-2189-minio-2                                     1/1     Running     0               19h     10.104.17.18    4am-node23   <none>           <none>
fouramf-bzvqc-73-2189-minio-3                                     1/1     Running     0               19h     10.104.23.164   4am-node27   <none>           <none>
fouramf-bzvqc-73-2189-pulsar-bookie-0                             1/1     Running     0               19h     10.104.6.62     4am-node13   <none>           <none>
fouramf-bzvqc-73-2189-pulsar-bookie-1                             1/1     Running     0               19h     10.104.1.59     4am-node10   <none>           <none>
fouramf-bzvqc-73-2189-pulsar-bookie-2                             1/1     Running     0               19h     10.104.24.73    4am-node29   <none>           <none>
fouramf-bzvqc-73-2189-pulsar-bookie-init-tn26t                    0/1     Completed   0               19h     10.104.1.50     4am-node10   <none>           <none>
fouramf-bzvqc-73-2189-pulsar-broker-0                             1/1     Running     0               19h     10.104.15.82    4am-node20   <none>           <none>
fouramf-bzvqc-73-2189-pulsar-proxy-0                              1/1     Running     0               19h     10.104.14.133   4am-node18   <none>           <none>
fouramf-bzvqc-73-2189-pulsar-pulsar-init-n2fdj                    0/1     Completed   0               19h     10.104.15.80    4am-node20   <none>           <none>
fouramf-bzvqc-73-2189-pulsar-recovery-0                           1/1     Running     0               19h     10.104.6.57     4am-node13   <none>           <none>
fouramf-bzvqc-73-2189-pulsar-zookeeper-0                          1/1     Running     0               19h     10.104.1.54     4am-node10   <none>           <none>
fouramf-bzvqc-73-2189-pulsar-zookeeper-1                          1/1     Running     0               19h     10.104.23.167   4am-node27   <none>           <none>
fouramf-bzvqc-73-2189-pulsar-zookeeper-2                          1/1     Running     0               19h     10.104.6.66     4am-node13   <none>           <none>
```

The monitoring shows that 18/98 indexing tasks have not been completed.
![image](https://github.com/milvus-io/milvus/assets/48523564/c1409eb0-d0ce-44cb-b31e-69209bc452f8)

**Comment by xiaocai2333** [comment]
@elstic please retest with latest image. 
pr was merged on 08/07.

**Comment by elstic** [comment]
> @elstic please retest with latest image. pr was merged on 08/07.

This problem has been fixed, verify image: 2.2.0-20230809-3322c5de
